### PR TITLE
Additional class paths for pr

### PR DESCRIPTION
**Pull request by andrewgilbertsagecom**
We had a need to write a custom PMD rule in java, and wanted to put our JAR alongside our source so CI can get it easily.

These changes add the workspace root to the classpath, and also adds a setting to be able to configure additional paths on the classpath, so the JAR can be found and the rule used.

### CODE REVIEW

**Comment by ChuckJonas** [comment]
@andrewgilbertsagecom thanks for submitting! 

Any chance you could share one of your custom rules for testing (or a just a `hello world` rule)?   I haven't gone through the process of writing a custom rule myself.

**Comment by ChuckJonas** [comment]
changed the target to `custom-ruleset-support` branch, which will also include an upgrade to the latest version of PMD (which allows defining of custom rules via `XPATH`

File: README.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ Allows you to run [Apex Static Analysis](http://pmd.sourceforge.net/snapshot/pmd
 - `priorityWarnThreshold`: Determines at what priority level 'warnings' will be added. Anything less will be a hint
 - `enableCache`: Creates a cache file for PMD to run faster. Will create a .pmdCache file in your workspace
 - `pmdBinPath` (prev. `pmdPath`) (optional): set to override the default pmd binaries. This should point to the PMD folder which contains folders `lib` and `bin`. Most likely it is called `libexec`.
+- `additionalClassPaths` (optional): set of paths to be appended to classpath. Used to find custom rule definitions. Can be absolute or relative to workspace.
```
Comment (ChuckJonas) [inline README.md]: We might want to include a link to the [PMD Documentation for how to define custom rules](https://pmd.github.io/latest/pmd_userdocs_extending_writing_pmd_rules.html)

Comment (ChuckJonas) [reply README.md]: > Used to find custom rule definitions 

Should this be:

Used to **define** custom rule definitions

File: src/lib/config.ts [hunk 2]
```diff
@@ -74,5 +76,19 @@ export class Config{
...
+            this.additionalClassPaths = [];
+        }
+
+        if (this.additionalClassPaths.length) {
+            this.additionalClassPaths = this.additionalClassPaths.map((unresolvedPath) => {
+                let resolvedPath = unresolvedPath
+                if (!path.isAbsolute(unresolvedPath) && vscode.workspace.rootPath) {
+                    resolvedPath = path.join(vscode.workspace.rootPath, unresolvedPath);
+                }
+                return resolvedPath
```
[the remark this answers is not in the document]
Comment (nchursin) [reply src/lib/config.ts]: @ChuckJonas we might want to add prettier and tslint to the repo

File: src/lib/apexPmd.ts [hunk 3]
```diff
@@ -139,7 +142,9 @@ export class ApexPmd {
 
         const pmdKeys = `${formatKey} ${cacheKey} ${targetPathKey} ${rulesetsKey}`
 
-        const cmd = `java -cp "${path.join(this._pmdPath, 'lib', '*')}" net.sourceforge.pmd.PMD ${pmdKeys}`;
+        const classPath = `${path.join(vscode.workspace.rootPath,'*')};${path.join(this._pmdPath, 'lib', '*')};${this._additionalClassPaths.join(';')}`
```
[the remark this answers is not in the document]
Comment (ChuckJonas) [reply src/lib/apexPmd.ts]: looks like we are currently using `vscode.workspace.rootPath` in 6 other locations 😬

Maybe we accept this PR with `vscode.workspace.rootPath` for now and create a followup issue to upgrade everything?

File: src/lib/config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChuckJonas) [reply src/lib/config.ts]: good idea... tracked in #48

File: README.md (hunk 1 above)
Comment (andrewgilbertsagecom) [reply README.md]: I've updated the readme and setting description.

It's not defining the custom rules, that is done in the java source, but rather is used to find the jar file containing those definitions.

File: src/lib/config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewgilbertsagecom) [reply src/lib/config.ts]: Oops, my bad. I've added in the missing `;`

File: src/lib/apexPmd.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nchursin) [reply src/lib/apexPmd.ts]: Oops 🙂 Sure, let's do that

**Comment by andrewgilbertsagecom** [comment]
I've put together a very simple sample rule [here](https://github.com/andrewgilbertsagecom/pmd-custom-rule-sample). You can see the source if you want to build it yourself, but I've also attached the jar if you just want to use it.

File: README.md (hunk 1 above)
Comment (ChuckJonas) [reply README.md]: @andrewgilbertsagecom  yea I understand.  Do you happen to have a simple jar that you could share?  Ideally we would have a simple "hello world" use case that we could use to make sure this feature works (maybe even unit test).

Comment (ChuckJonas) [reply README.md]: Disregard that last comment! should have kept reading! 

Going to try and run it with your simple rule.  Thanks!

**Comment by ChuckJonas** [comment]
@andrewgilbertsagecom can you help me understand why we add the workspace root to the class path? 

When I clone down this PR and run it I get 
>Error: Could not find or load main class net.sourceforge.pmd.PMD

If I remove the `${path.join(vscode.workspace.rootPath,'*')};` it works (because `cmd` is now essentially the same as it was before).

Haven't tried to actually include your jar yet...

**Comment by ChuckJonas** [comment]
Same thing if I copy down your jar and then add it via the new `additionalClassPaths` setting...

I'm not sure what the issue is.. I've definitely regressed to java newb these days

**Comment by ChuckJonas** [comment]
Think I figured it out... I would be money your a windows user 😉 

Looks like we have to delimitate the cp different on windows (`;`) vs osx/linux (`:`).

Once I switched to `:` (and upgraded my java runtime) I was able to get it to work!  I'm going to go ahead and merge this PR and then I'll submit a fix

**Pull request merged**